### PR TITLE
Select (dropdown) support on the web frontend

### PR DESCRIPTION
**Pull request by ThoAppelsin**
## About

* **Related Issues:** #167 (done), #168 (undone), #173 **(will be done when approved)**
* **Other Related Links:** [Documentation for the JSON Structure of Data Template Fields and Posts](../wiki/Documentation-for-the-JSON-Structure-of-Data-Template-Fields-and-Posts)

To give our users an interface to _select one out of many options_ kind of input, we needed either a group of radio buttons, or a dropdown. I have made the necessary changes to support the select (dropdown) fields on the web frontend.

Its options (items) can be edited similar to the way that the labels are editable. This time, with an auto-resizing (in height and width) textarea.

**Note:** The branch is out-of-date. The changes made, however, should be isolated. Tell me to merge the master into it, if necessary.

### Changes

- Introduction of the Select (friendly name: Dropdown) input and field on the templates.
- Ability to edit the select (dropdown) input, create a template with it, share posts using it, display posts using them.
  - Clicking away from the editing box solidifies the changes.
  - Each row is for an item on the dropdown.
  - Hitting the key combination `Control+Enter` also finalizes the edit.
  - Deleting everything from the content reverts the dropdown back to how it was before the edit started.
  - Blank lines are excluded.
- The order of available fields on the template creation screen is now alphabetical with respect to the friendly names, for easier access.

## Visuals

dropdowns

Note 1: The animation fails to depict the horizontal auto-resizing of the editing area.  
Note 2: The branch is out-of-date with respect to the master branch, which got two more recent merges, and so is the animation.

## Testing

How can this contribution be tested? Use the following structure:

- Create a template using the new Dropdown field.
  - Edit the contents of the Dropdown field.
  - Try editing it into nothingness, observe whether it reverts back to its initial state before the edit.
  - Try expanding the box in both dimensions, try finding a flaw.
  - Try making blank select options, shouldn't be possible.
  - Try tricking the edit box into yielding an empty dropdown.
- Share a post using the new template.
  - Observe if the shared post is as expected.

### CODE REVIEW

File: api/interestr/website/templates/website/group-details.html [hunk 1]
```diff
@@ -481,43 +477,156 @@ <h5 class="modal-title" id="createTemplateModalLabel">Create a template</h5>
...
+        var select = makeSelect((selectEditorTextarea.textContent || selectEditorInitial).split('\n'));
+        selectEditorTextarea.parentNode.replaceChild(select, selectEditorTextarea);
+
+        selectEditorTextarea = null;
+
+        modal.focus();
+      }
+    }
+
+    // Credits: https://stackoverflow.com/a/21015393
```
Comment (smddzcy) [inline api/interestr/website/templates/website/group-details.html:492]: lol

File: api/interestr/website/templates/website/group-details.html [hunk 2]
```diff
@@ -656,74 +764,67 @@ <h5 class="modal-title" id="createTemplateModalLabel">Create a template</h5>
...
 
-    for (const type in templateFieldTypes) {
+    function templateFieldTypesSorted(accordingTo) {
+      return Object.getOwnPropertyNames(templateFieldTypes).sort((a, b) =>
+        templateFieldTypes[a][accordingTo].localeCompare(templateFieldTypes[b][accordingTo])
+      );
+    }
+
+    // for (const type in templateFieldTypes) {
+    for (const type of templateFieldTypesSorted('friendlyName')) {
```
Comment (smddzcy) [inline api/interestr/website/templates/website/group-details.html:827]: Cool, better UX 👌

**Review by smddzcy — APPROVED** [review]
Super, :shipit: if it works.

File: api/interestr/website/templates/website/group-details.html [hunk 3]
```diff
@@ -481,43 +477,156 @@ <h5 class="modal-title" id="createTemplateModalLabel">Create a template</h5>
       }
     }
 
+    function replaceTextareaWithSelect() {
+      if (selectEditorTextarea) {
+        selectEditorTextarea.textContent = selectEditorTextarea.textContent.trim();
```
Comment (Morgazipa) [inline api/interestr/website/templates/website/group-details.html:482]: To be able to update the options in the dropdown I had to change the RHS to "selectEditorTextarea.value.trim();".  Again, I was testing in Firefox version 56.0.2.

**Comment by Morgazipa** [comment]
Unfortunately,  due to changes made to serializers in PR #171,  creation of posts/comments does not work. Although I tried to fix the problems introduced in #172, I only changed data template serializer because I thought that the tests for comments  and posts would pass if we would re-write them using "data" field instead of "text" in post creation.

**Comment by smddzcy** [comment]
@Morgazipa @triforce930 What is the final status; do posting/commenting work now, and can we merge this PR?

**Comment by Morgazipa** [comment]
@smddzcy Unfortunately, not yet.

**Comment by smddzcy** [comment]
@triforce930 We have 2 options here:
1. Reverting #171 
2. Fixing #171 

Which one do you choose 😄

**Comment by karacasoft** [comment]
I'll fix it today :+1:. Sorry for the inconvenience

**Comment by smddzcy** [comment]
Since everything is fixed, we can merge it now.

**Pull request merged**